ert_gui/tools/file/file_model.py

```diff
@@ -0,0 +1,49 @@
+from ErtQt.Qt import pyqtSlot, Qt, QAbstractListModel, QModelIndex, QApplication, QClipboard
+
+
+class FileModel(QAbstractListModel):
+    def __init__(self, parent=None):
+        super(FileModel, self).__init__(parent)
+        self._rows = []
+        self._new_line = True
+
+    def rowCount(self, parent):
+        """
+        Overloaded Qt function. Return the number of rows in this model.
+        :type parent: QModelIndex
+        """
+        return len(self._rows)
+
+    def data(self, index, role=Qt.DisplayRole):
+        """
+        Overloaded Qt function. Return data for index.
+        :type index: QModelIndex
+        """
+        if not index.isValid() or index.row() >= len(self._rows):
+            return
+
+        line = self._rows[index.row()]
+        if role == Qt.DisplayRole:
+            return line
+
+    @pyqtSlot(str)
+    def append_text(self, text):
+        if type(text).__name__ == "QString":
+            # PyQt4 sens QString instead of str, so convert it.
+            text = str(text)
+        if not self._new_line:
+            text = self._rows[-1] + text
+        self._new_line = text[-2] == "\n"
+
+        rows = text.splitlines(False)
+        first = len(self._rows)
+        last = first + len(rows)
+
+        self.beginInsertRows(QModelIndex(), first, last)
+        self._rows.extend(rows)
+        self.endInsertRows()
```

Comment:
don't quite follow of why there is begin and end insert

Reply:
Qt emits a bunch of signals before/after insertion. In particular, we're interested in the `rowsInserted` signal. It might also do some thread stuff so there's no race conditions (maybe).

https://doc.qt.io/qt-5/qabstractitemmodel.html#beginInsertRows